File: 2069-walking-robot-simulation-ii/2069-walking-robot-simulation-ii.go
var dirs = []string{"North", "East", "South", "West"}
var steps = [][]int{{0, 1}, {1, 0}, {0, -1}, {-1, 0}}

type Robot struct {
	w int
	h int
	d int // 0: north, 1: east, 2: south, 3: west
	p []int
}

func Constructor(width int, height int) Robot {
	return Robot{
		w: width,
		h: height,
		d: 1,
		p: []int{0, 0},
	}
}

func (this *Robot) Step(num int) {
    cycle := 2 * (this.w + this.h) - 4
    num = num % cycle

    // edge case: num % cycle == 0 when at pos (0, 0)
    // return to (0, 0), but the dir must be handled
    if num == 0 {
        if this.p[0] == 0 && this.p[1] == 0 {
            this.d = 2 // south
        }
        return
    }

    for i := 0; i < num; i++ {
        nx, ny := this.p[0] + steps[this.d][0], this.p[1] + steps[this.d][1]
        if nx < 0 || nx >= this.w || ny < 0 || ny >= this.h {
            this.d = (this.d - 1 + 4) % 4
            nx, ny = this.p[0] + steps[this.d][0], this.p[1] + steps[this.d][1]            
        }
        this.p[0], this.p[1] = nx, ny
    }
}

func (this *Robot) GetPos() []int {
	return this.p
}

func (this *Robot) GetDir() string {
	return dirs[this.d]
}

// design, simulation
// time: O(n)
// space: O(1)

// note:
// crazy edge case

/**
 * Your Robot object will be instantiated and called as such:
 * obj := Constructor(width, height);
 * obj.Step(num);
 * param_2 := obj.GetPos();
 * param_3 := obj.GetDir();
 */

File: 2069-walking-robot-simulation-ii/2069-walking-robot-simulation-ii.py
class Robot:
    dirs = ["North", "East", "South", "West"]
    steps = [(0, 1), (1, 0), (0, -1), (-1, 0)]

    def __init__(self, width: int, height: int):
        self.w = width
        self.h = height
        self.d = 1
        self.p = [0, 0]

    def step(self, num: int) -> None:
        cycle = 2 * (self.w + self.h) - 4
        num = num % cycle

        # edge case
        if num == 0:
            if self.p[0] == 0 and self.p[1] == 0:
                self.d = 2 # south
                return

        for i in range(num):
            nx, ny = self.p[0] + self.steps[self.d][0], self.p[1] + self.steps[self.d][1]
            if nx < 0 or nx >= self.w or ny < 0 or ny >= self.h:
                self.d = (self.d - 1 + 4) % 4
                nx, ny = self.p[0] + self.steps[self.d][0], self.p[1] + self.steps[self.d][1]
            self.p[0], self.p[1] = nx, ny

    def getPos(self) -> List[int]:
        return self.p

    def getDir(self) -> str:
        return self.dirs[self.d]
        
# design, simulation
# time: O(n)
# space: O(1)

# Your Robot object will be instantiated and called as such:
# obj = Robot(width, height)
# obj.step(num)
# param_2 = obj.getPos()
# param_3 = obj.getDir()